### PR TITLE
writer: discard closed connections in pub

### PR DESCRIPTION
**Pull request by panaetov**
Writer publishes messages on opened connections only.

### CODE REVIEW

File: nsq/writer.py [hunk 1]
```diff
@@ -128,11 +128,14 @@ def _pub(self, command, topic, msg, delay_ms=None, callback=None):
...
-            callback(None, protocol.SendError('no connections'))
-            return
-
-        conn = random.choice(list(self.conns.values()))
+        open_connects = [
+            conn for conn in self.conns.values()
+            if conn.connected()
+        ]
+        if not open_connects:
+            raise PubFailure("Not open connections")
```
Comment (mreiferson) [inline nsq/writer.py]: I don't want to introduce different failure semantics (exceptions) — let's use:

```python
callback(None, protocol.SendError('no open connections'))
```

File: nsq/writer.py [hunk 2]
```diff
@@ -128,11 +128,14 @@ def _pub(self, command, topic, msg, delay_ms=None, callback=None):
             callback = functools.partial(self._finish_pub, command=command,
                                          topic=topic, msg=msg)
 
-        if not self.conns:
-            callback(None, protocol.SendError('no connections'))
-            return
-
-        conn = random.choice(list(self.conns.values()))
+        open_connects = [
```
Comment (mreiferson) [inline nsq/writer.py]: `s/open_connects/open_connections`

File: nsq/writer.py [hunk 3]
```diff
@@ -143,9 +146,9 @@ def _pub(self, command, topic, msg, delay_ms=None, callback=None):
 
         try:
             conn.send(cmd(*args))
-        except Exception:
-            logger.exception('[%s] failed to send %s' % (conn.id, command))
+        except:
             conn.close()
+            raise PubFailure('[%s] failed to send %s' % (conn.id, command))
```
Comment (mreiferson) [inline nsq/writer.py]: same here, don't want to introduce new failure semantics, let's drop the `raise` please

File: nsq/writer.py (hunk 1 above)
Comment (panaetov) [inline nsq/writer.py]: Done

File: nsq/writer.py (hunk 3 above)
Comment (panaetov) [inline nsq/writer.py]: Done

File: nsq/writer.py (hunk 2 above)
Comment (panaetov) [inline nsq/writer.py]: Done

File: nsq/writer.py [hunk 4]
```diff
@@ -143,8 +147,9 @@ def _pub(self, command, topic, msg, delay_ms=None, callback=None):
 
         try:
             conn.send(cmd(*args))
-        except Exception:
+        except:
```
Comment (ploxiln) [inline nsq/writer.py]: Why this change?

It's generally a good idea to catch no more broadly than `Exception`. There are a couple exceptions which intentionally do not subclass it, like `SystemExit` and `KeyboardInterrupt`, because catching those can (intermittently) break exiting and restarting the process.

Comment (panaetov) [inline nsq/writer.py]: Thanks, done.

File: nsq/writer.py [hunk 5]
```diff
@@ -145,6 +149,7 @@ def _pub(self, command, topic, msg, delay_ms=None, callback=None):
             conn.send(cmd(*args))
         except Exception:
             logger.exception('[%s] failed to send %s' % (conn.id, command))
+            callback(None, protocol.SendError('sending error'))
```
Comment (mreiferson) [inline nsq/writer.py]: I'm not 100% sure of this change — what issue is this addressing?

Comment (ploxiln) [inline nsq/writer.py]: I think that without this, the caller is not informed that it possibly failed to send the message (even after seemingly having a connection).

Comment (mreiferson) [inline nsq/writer.py]: Ahh, right. In that case `s/sending error/send error` please :)

Comment (panaetov) [inline nsq/writer.py]: Done

**Comment by mreiferson** [comment]
LGTM, thanks for sticking with this! Would you mind squashing down to one commit?

**Comment by protoss-player** [comment]
Hi @mreiferson,

I believe you can squash everything down when merging this PR. As per https://github.com/blog/2141-squash-your-commits, there should be "Confirm squash and merge" option.

**Comment by mreiferson** [comment]
Yea, I know, I actually prefer the merge commit for historical reasons, and that option isn't offered in GH.

**Comment by panaetov** [comment]
Done. Everything is squashed down to one commit.

**Comment by mreiferson** [comment]
Thanks!
